modin/backends/pandas/query_compiler.py

```diff
@@ -561,21 +561,48 @@ def reset_index(self, **kwargs):
         """
         drop = kwargs.get("drop", False)
         level = kwargs.get("level", None)
-        # TODO Implement level
-        if level is not None or self.has_multiindex():
-            return self.default_to_pandas(pandas.DataFrame.reset_index, **kwargs)
+        new_index = None
         if not drop:
-            new_column_name = (
-                self.index.name
-                if self.index.name is not None
-                else "index"
-                if "index" not in self.columns
-                else "level_0"
-            )
-            new_self = self.insert(0, new_column_name, self.index)
+            if self.has_multiindex():
+                col_level = kwargs.get("col_level", 0)
+                col_fill = kwargs.get("col_fill", "")
+                # Convert multiindex into a dataframe with columns containing multiindex levels
+                mi_frame = self.index.to_frame()
+                if level is not None:
+                    if not isinstance(level, (tuple, list)):
+                        level = [level]
+                    level = [self.index._get_level_number(lev) for lev in level]
+                    uniq_sorted_level = sorted(set(level))
+                    if len(level) < self.index.nlevels:
+                        new_index = self.index.droplevel(uniq_sorted_level)
+                    mi_frame = mi_frame.iloc[:, uniq_sorted_level]
+                if self.columns.nlevels > 1 and mi_frame.columns.nlevels == 1:
```

Comment:
condition should be something like:
```python
if self.columns.nlevels > mi_frame.columns.nlevels
```
 and the logic in this if-branch should also be changed, because it's going to processes situations like this incorrectly:
```python
>>> df
                           a     b
                           0  1  0  1
                           z  z  z  z
(level-0, 0) (level-0, 1)
a            0             0  0  0  0
             1             1  1  1  1
b            0             2  2  2  2
             1             3  3  3  3
             2             4  4  4  4
>>> df.index
MultiIndex([('a', 0),
            ('a', 1),
            ('b', 0),
            ('b', 1),
            ('b', 2)],
           names=[('level-0', 0), ('level-0', 1)])
>>> df.columns
MultiIndex([('a', 0, 'z'),
            ('a', 1, 'z'),
            ('b', 0, 'z'),
            ('b', 1, 'z')],
           )
>>> df.reset_index(drop=False)
  level-0     a     b
        0  1  0  1  0  1
              z  z  z  z
0       a  0  0  0  0  0
1       a  1  1  1  1  1
2       b  0  2  2  2  2
3       b  1  3  3  3  3
4       b  2  4  4  4  4
```

Reply:
I updated tests and functions with tuples in level names. Number of test cases only grew up because of it, but I think that all parameter combinations are necessary.